### PR TITLE
Added checks for dice_random

### PR DESCRIPTION
**Pull request by soundarya98**
In response to #161 
- Added check to throw an error if query instance has an unknown column
- Added tests for dice_genetic, dice_KD to check if query instance has an unknown column
- Fixed bug in helpers.py - added parameter invalid_raise = True for the adult dataset
- Fixed bug in tests: query instance had categorical values not present in training data
- Added tests for dice_random

### CODE REVIEW

File: dice_ml/explainer_interfaces/dice_random.py [hunk 1]
```diff
@@ -138,7 +132,13 @@ class of query_instance for binary classification.
             if len(cfs_df) > total_CFs:
                 cfs_df = cfs_df.sample(total_CFs)
             cfs_df.reset_index(inplace=True, drop=True)
-            self.cfs_pred_scores = self.predict_fn(cfs_df)
+            if len(cfs_df) > 0:
+                self.cfs_pred_scores = self.predict_fn(cfs_df)
+            else:
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_random.py:137]: seems arbitrary to assign zero to the `cfs_pred_scores`. Is None a better alternative?

File: tests/test_dice_interface/test_dice_random.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                                                 features_to_vary=features_to_vary,
+                                                 total_CFs=total_CFs, desired_class=desired_class,
+                                                 desired_range=desired_range, permitted_range=permitted_range)
+
+        for feature in self.exp.data_interface.feature_names:
+            if feature not in features_to_vary:
+                assert all(ans.final_cfs_df[feature].values[i] == sample_custom_query_2[feature].values[0] for i in
+                           range(total_CFs))
+
+    # # Testing that the permitted_range argument actually varies the features only within the permitted_range
```
Comment (amit-sharma) [inline tests/test_dice_interface/test_dice_random.py:91]: can these commented lines be removed?

**Review by amit-sharma — CHANGES_REQUESTED** [review]
Added two comments to review

File: tests/test_dice_interface/test_dice_random.py (hunk 2 above)
Comment (soundarya98) [reply tests/test_dice_interface/test_dice_random.py:91]: I left this check commented as dice_random currently doesn't implement permitted_range properly, in the future if that changes, we can un-comment the test

**Comment by soundarya98** [comment]
> Added two comments to review

@amit-sharma, the if statement in line 135 is not redundant. Even though we are checking that `len(cfs_df)>0` in line 131, in line 133, we are sampling `total_CFs` number of CFs. When `total_CFs = 0`, `cfs_df` now becomes 0, and the predict step would throw an error without the check

File: tests/test_dice_interface/test_dice_random.py (hunk 2 above)
Comment (amit-sharma) [reply tests/test_dice_interface/test_dice_random.py:91]: okay

**Review by amit-sharma — APPROVED** [review]

**Comment by amit-sharma** [comment]
@gaugup does this look good?

**Comment by gaugup** [comment]
> @gaugup does this look good?

Sorry, it slipped through. I will take a look in my AM time.

File: tests/test_dice_interface/test_dice_random.py (hunk 2 above)
Comment (gaugup) [reply tests/test_dice_interface/test_dice_random.py:91]: Can you mark the test skipped and document the reason why this test is skipped or remove the test case altogether? We shouldn't keep commented out code.

File: dice_ml/utils/helpers.py [hunk 3]
```diff
@@ -20,7 +20,7 @@ def load_adult_income_dataset(only_train=True):
     :return adult_data: returns preprocessed adult income dataset.
     """
     raw_data = np.genfromtxt('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data',
-                             delimiter=', ', dtype=str)
+                             delimiter=', ', dtype=str, invalid_raise=False)
```
Comment (gaugup) [inline dice_ml/utils/helpers.py:23]: Why is the invalid_raise set to False here? It seems if set to True it correctly raises the exception.

Comment (soundarya98) [reply dice_ml/utils/helpers.py:23]: One of the adult dataset rows has an extra space or something like that, which makes the `genfromtext` function think that there is an extra column. This was raising an error. `invalid_raise` just skips that problematic row

Comment (gaugup) [reply dice_ml/utils/helpers.py:23]: Got it. Thanks for clarifying.

**Review by gaugup — APPROVED** [review]

**Pull request merged**